compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -142,27 +142,25 @@ pub fn relate_substs<'tcx, R: TypeRelation<'tcx>>(
     b_subst: SubstsRef<'tcx>,
 ) -> RelateResult<'tcx, SubstsRef<'tcx>> {
     let tcx = relation.tcx();
-    let mut cached_ty = None;
 
-    let params = iter::zip(a_subst, b_subst).enumerate().map(|(i, (a, b))| {
-        let (variance, variance_info) = match variances {
-            Some((ty_def_id, variances)) => {
+    let zipped = iter::zip(a_subst, b_subst);
+    match variances {
+        Some((ty_def_id, variances)) => {
+            let ty = tcx.type_of(ty_def_id).subst(tcx, a_subst);
+            tcx.mk_substs(zipped.enumerate().map(|(i, (a, b))| {
                 let variance = variances[i];
                 let variance_info = if variance == ty::Invariant {
-                    let ty = *cached_ty
```

Comment:
Was this cache just not very effective?

Reply:
Do you know the relative frequencies of the `Some` case and the `None` case?

Reply:
It's more that the cahce doesn't do anything now that I've reordered stuff a bit, in the `Some` case you just look up the type once, and iterate after that. Unless I'm missing something in how this worked?